lib/jxl/enc_ac_strategy.cc

```diff
@@ -659,6 +713,149 @@ void FindBest16X16(size_t bx, size_t by, size_t cx, size_t cy,
+      MultiBlockTransformCrossesVerticalBoundary(*ac_strategy, bx + cx + 4,
+                                                 by + cy, by + cy + 4)) {
+    return;  // not suitable for 32x32 analysis, some transforms leak out.
+  }
+  const bool allow_32X16 = !MultiBlockTransformCrossesVerticalBoundary(
+      *ac_strategy, bx + cx + 2, by + cy, by + cy + 4);
+  const bool allow_16X32 = !MultiBlockTransformCrossesHorizontalBoundary(
+      *ac_strategy, bx + cx, by + cy + 2, bx + cx + 4);
+  // Current entropies aggregated on 16x16 resolution.
+  float entropy[2][2] = {};
+  for (size_t dy = 0; dy < 4; ++dy) {
+    for (size_t dx = 0; dx < 4; ++dx) {
+      entropy[dy >> 1][dx >> 1] += entropy_estimate[(cy + dy) * 8 + (cx + dx)];
+    }
+  }
+  float try32X16_0 = std::numeric_limits<float>::max();
+  float try32X16_1 = std::numeric_limits<float>::max();
+  float try16X32_0 = std::numeric_limits<float>::max();
+  float try16X32_1 = std::numeric_limits<float>::max();
+  if (allow_32X16) {
+    if (row0[bx + cx + 0].RawStrategy() != acs_raw32X16) {
+      try32X16_0 =
+          entropy_mul * EstimateEntropy(acs32X16, (bx + cx + 0) * 8,
+                                        (by + cy + 0) * 8, config, cmap_factors,
+                                        block, scratch_space, quantized);
+    }
+    if (row0[bx + cx + 2].RawStrategy() != acs_raw32X16) {
+      try32X16_1 =
+          entropy_mul * EstimateEntropy(acs32X16, (bx + cx + 2) * 8,
+                                        (by + cy + 0) * 8, config, cmap_factors,
+                                        block, scratch_space, quantized);
+    }
+  }
+  if (allow_16X32) {
+    if (row0[bx + cx].RawStrategy() != acs_raw16X32) {
+      try16X32_0 =
+          entropy_mul * EstimateEntropy(acs16X32, (bx + cx + 0) * 8,
+                                        (by + cy + 0) * 8, config, cmap_factors,
+                                        block, scratch_space, quantized);
+    }
+    if (row1[bx + cx].RawStrategy() != acs_raw16X32) {
+      try16X32_1 =
+          entropy_mul * EstimateEntropy(acs16X32, (bx + cx + 0) * 8,
+                                        (by + cy + 2) * 8, config, cmap_factors,
+                                        block, scratch_space, quantized);
+    }
+  }
+  float try32X32 =
+      entropy_mul_32X32 *
+      EstimateEntropy(acs32X32, (bx + cx + 0) * 8, (by + cy + 0) * 8, config,
+                      cmap_factors, block, scratch_space, quantized);
+
+  // Test if this block should have 32X16 or 16X32 transforms,
+  // because it can have only one or the other.
+  float cost32x16 = std::min(try32X16_0, entropy[0][0] + entropy[1][0]) +
+                    std::min(try32X16_1, entropy[0][1] + entropy[1][1]);
+  float cost16x32 = std::min(try16X32_0, entropy[0][0] + entropy[0][1]) +
+                    std::min(try16X32_1, entropy[1][0] + entropy[1][1]);
+  if (try32X32 < cost32x16 && try32X32 < cost16x32) {
```

Comment:
It sounds to me that this whole if structure should be refactored to not have basically the same code repeated 4 times (and I assume that the 16x16 function is very similar too)

Reply:
I'll generalize this in a next step to 32X64,64X32 and 64X64, and then see how to simplify it.

I'll try to keep refactorings and new development in separate PRs so that there is more clarity of no adverse effects from refactorings.

Reply:
Adding a TODO about this

Reply:
Fair enough, as long as the final state doesn't have this repetition.

Note that the argument of "no adverse effects from refactorings" doesn't apply to not repeating the code 4 times with slightly different constants in the lines below, though - it would be better to make those parts a lambda.

Reply:
changed my mind and did a small refactoring already, but think that more is needed -- however that I'd like to postpone to the point when this code is first generalized to all sizes